examples_tests/43.AllocatorTest/main.cpp

```diff
@@ -1,53 +1,219 @@
 #define _IRR_STATIC_LIB_
-
 #include <irrlicht.h>
-#include "irr/core/alloc/GeneralpurposeAddressAllocator.h"
-
-using namespace irr;
+#include <random>
+#include <cmath>
 
+using namespace irr;
+using namespace core;
 
-/*
-	Problems with GeneralPurposeallocator:
-		1. GeneralPurposeallocator is not able to allocate as much address space as it is expected to
-		2. Program doesn't compile, when I try to call safe_shrink_size
-		3. "Assertion failed: false, file *\include\irr\core\alloc\GeneralpurposeAddressAllocator.h, line 368"
+/*
+	If traits say address allocator supports arbitrary order free,
+	then ETAT_CHOOSE_RANDOMLY is valid. Otherwise ETAT_CHOOSE_MOST_RECENTLY_ALLOCATED
 */
-
-int main()
+
+enum E_TRAITS_ALLOCATION_TYPE
+{
+	ETAT_CHOOSE_RANDOMLY,
+	ETAT_CHOOSE_MOST_RECENTLY_ALLOCATED,
+	ETAT_COUNT
+};
+
+struct AllocationCreationParameters
+{	
+	size_t multiAllocCnt;              //! Specifies amount of adress to be allocated with certain choosen allocator
+	size_t adressesToDeallocateCnt;    //! Specifies amount of adress to be deallocated with certain choosen allocator. Must be less than all allocated and must pick number less than `traits::max_multi_free`, but we don't have `max_multi_free`
+};
+
+constexpr size_t minAllocs = 10000u;
+constexpr size_t maxAllocs = 20000u;
+constexpr size_t maxSizeForSquaring = 1024u;
+constexpr size_t maxAlignmentExp = 6u;
+constexpr size_t minByteSizeForReservingMemory = 2048;			// 2kB
+constexpr size_t maxByteSizeForReservingMemory = 2147483648;	// 2GB
+constexpr size_t maxOffset = 512;
+constexpr size_t maxAlignOffset = 64;
+constexpr size_t maxBlockSize = 600;
+
+//TODO: full static
+class RandomNumberGenerator
+{
+public:
+	inline uint32_t getRndAllocCnt()    { return  allocsPerFrameRange(mt);  }
+	inline uint32_t getRndSize()        { return  sizePerFrameRange(mt); }
+	inline uint32_t getRndMaxAlign()    { return  (1u << maxAlignmentExpPerFrameRange(mt)); } //128 is max
```

Comment:
max alignment and size do not really apply to the Pool Address allocator, because there the allocations must be a fixed size (block size) so the allocation and size will be both equal to that.

You can however, test different block sizes for multiple instances of Pool Address Allocators.

Reply:
Also I don't think there's really a requirement for my Address Allocators to only service PoT alignments, in-fact the alignments can be any non-zero positive integer.... this is why GeneralPurposeAddressAllocator can be used to manage any format mesh vertices ;)

Reply:
> Also I don't think there's really a requirement for my Address Allocators to only service PoT alignments
https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/include/irr/core/alloc/AddressAllocatorBase.h#L73

Reply:
that's max requestable that needs to be PoT, not the individuals.